src/main/scala/io/opentargets/etl/backend/Expression.scala

```diff
@@ -1,33 +1,288 @@
+      .withColumnRenamed("_c1", "name")
+
+    val mapEfosRenamed = mapEfos.withColumnRenamed("tissue_id", "tissue_internal_id")
+
+    val mappedInfo = mapEfosRenamed
+      .join(expressionsRenamed, col("name") === col("tissue_internal_id"), "full")
+      .withColumn("efoId", when(col("efo_code").isNull, col("name")).otherwise(col("efo_code")))
+      .withColumn("labelNew", when(col("label").isNull, col("name")).otherwise(col("label")))
+
+    mappedInfo
+  }
+
+  /*
+     It generates the list of valid gene and relative labels.
+   */
+  private def selectTissues(tissues: DataFrame, efoTissueMap: DataFrame) = {
+    val normalTissueLabel =
+      tissues.join(efoTissueMap, col("labelNew") === col("Tissue"), "left")
+    val normalTissueExpression =
+      tissues.join(efoTissueMap, col("expressionId") === col("Tissue"), "left")
+
+    val normalTissueWithLabel = normalTissueLabel.unionByName(normalTissueExpression)
+
+    val emptyLabels = normalTissueWithLabel
+      .filter(col("labelNew").isNull)
+      .withColumn("TissueDef", col("Tissue"))
+      .select("Gene", "TissueDef")
+
+    val hasLabels = normalTissueWithLabel
+      .filter(col("labelNew").isNotNull)
+
+    val hasLabelFiltered = hasLabels
+      .withColumn("TissueDef", col("Tissue"))
+      .select("Gene", "TissueDef")
+
+    val missingLabel = emptyLabels.except(hasLabelFiltered)
+
+    val selectMissingRecords = normalTissueWithLabel
+      .join(missingLabel, Seq("Gene"), "right")
+      .where(col("TissueDef") === col("Tissue"))
+      .withColumn("labelDef", col("TissueDef"))
+      .drop("TissueDef")
+
+    val validLabels = hasLabels
+      .withColumn("labelDef", col("LabelNew"))
+      .unionByName(selectMissingRecords)
+
+    validLabels
+  }
+
+  /*
+    Given the list of gene, tissues it fills the info with rna,level,rscore and unit info.
+   */
+  private def generateBaselineInfo(normalTissueDF: DataFrame,
+                                   baselineExpressionDF: DataFrame): DataFrame = {
+    val normalTissueKeyDF =
+      normalTissueDF
+        .withColumn("key", concat(col("Gene"), lit('-'), col("Tissue")))
+        .withColumnRenamed("Gene", "GeneNormal")
+        .withColumnRenamed("Tissue", "TissueNormal")
+
+    val baselineExpressionKeyDF =
+      baselineExpressionDF
+        .withColumn("key", concat(col("Gene"), lit('-'), col("Tissue")))
+        .withColumnRenamed("Gene", "GeneBase")
+        .withColumnRenamed("Tissue", "TissueBase")
+
+    val unionByKey = normalTissueKeyDF.join(baselineExpressionKeyDF, Seq("key"), "full")
+
+    val tissueBaselineInfoDF = unionByKey
+      .withColumn("Gene",
+                  when(col("GeneNormal").isNull, col("GeneBase")).otherwise(col("GeneNormal")))
+      .withColumn(
+        "Tissue",
+        when(col("TissueNormal").isNull, col("TissueBase")).otherwise(col("TissueNormal")))
+      .withColumn("LevelMapDef", when(col("LevelMap").isNull, -1).otherwise(col("LevelMap")))
```

Comment:
These `withColumn` expressions can generally be formulated as df.select(coalesce(col("LevelMap"), lit(-1)) as "LevelMapDef"). 

This has the advantage of being able to put all of these checks in a single expression and makes clear that the default is in the column and the literal is the fallback. 

From the `withColumn` documentation it also warns about potential performance problems caused by chaining too many of these clauses together: 

> Returns a new Dataset by adding a column or replacing the existing column that has the same name.
> column's expression must only refer to attributes supplied by this Dataset. It is an error to add a column that refers to some other Dataset.
> Note:
> this method introduces a projection internally. Therefore, calling it multiple times, for instance, via loops in order to add multiple columns can generate big plans which can cause performance issues and even StackOverflowException. To avoid this, use select with the multiple columns at once.

Reply:
This is a really good point.
Agree on the fact that withColumn has a cost vs select/coalesce approach.
I am changing the code and re-run everything. (double check the output due the migration from data-pipeline)